definition_validator/validator.cc

```diff
@@ -189,6 +189,78 @@ void validateOverriding(const core::GlobalState &gs, core::SymbolRef method) {
+    }
+    for (const auto &anc : classDef->singletonAncestors) {
+        const auto ancConst = ast::cast_tree<ast::ConstantLit>(anc.get());
+        if (ancConst == nullptr) {
+            continue;
+        }
+        if (ancConst->symbol.data(gs)->dealias(gs) == ancestor) {
+            return anc->loc;
+        }
+    }
+    // give up
+    return classDef->loc;
+}
+
+void validateFinalAncestorHelper(const core::GlobalState &gs, const core::SymbolData classData,
+                                 const ast::ClassDef *const classDef, const core::SymbolData errMsgClassData,
+                                 const std::string verb) {
+    for (const auto &mixin : classData->mixins()) {
+        if (!mixin.data(gs)->isClassFinal()) {
+            continue;
+        }
+        if (auto e = gs.beginError(getAncestorLoc(gs, classDef, mixin), core::errors::Resolver::FinalAncestor)) {
+            e.setHeader("`{}` was declared as final and cannot be {} by `{}`", mixin.data(gs)->show(gs), verb,
+                        errMsgClassData->show(gs));
+            e.addErrorLine(mixin.data(gs)->loc(), "`{}` defined here", mixin.data(gs)->show(gs));
+        }
+    }
+}
+
+void validateFinalMethodHelper(const core::GlobalState &gs, const core::SymbolData classData,
+                               const core::SymbolData errMsgClassData) {
+    if (!classData->isClassFinal()) {
+        return;
+    }
+    for (const auto mem : classData->members()) {
+        const auto memData = mem.second.data(gs);
+        if (!memData->isMethod() || memData->name == core::Names::staticInit() || memData->isFinalMethod()) {
+            continue;
+        }
+        if (auto e = gs.beginError(memData->loc(), core::errors::Resolver::FinalModuleNonFinalMethod)) {
+            e.setHeader("`{}` was declared as final but its method `{}` was not declared as final",
+                        errMsgClassData->show(gs), memData->name.show(gs));
+        }
+    }
+}
+
+void validateFinal(const core::GlobalState &gs, const core::SymbolData classData, const ast::ClassDef *const classDef) {
+    const auto superClass = classData->superClass();
+    if (superClass.exists() && superClass.data(gs)->isClassFinal()) {
+        if (auto e = gs.beginError(getAncestorLoc(gs, classDef, superClass), core::errors::Resolver::FinalAncestor)) {
+            e.setHeader("`{}` was declared as final and cannot be inherited by `{}`", superClass.data(gs)->show(gs),
+                        classData->show(gs));
+            e.addErrorLine(superClass.data(gs)->loc(), "`{}` defined here", superClass.data(gs)->show(gs));
+        }
+    }
+    validateFinalAncestorHelper(gs, classData, classDef, classData, "included");
+    validateFinalMethodHelper(gs, classData, classData);
+    const auto singletonData = classData->lookupSingletonClass(gs).data(gs);
+    validateFinalAncestorHelper(gs, singletonData, classDef, classData, "extended");
+    validateFinalMethodHelper(gs, singletonData, classData);
```

Comment:
I appreciate that you're going to a lot of care to avoid calling `->data()` on symbols, but this is really foreign-looking code. Why can't we just only pass around the class symbol, and compute what we need when we need it? 
This incidentally makes the code easier to read because the reader doesn't have to know where some symbol data came from; they saw it computed from another symbol right there.

Reply:
I guess at a higher level, it seems like you're really trying to not repeat yourself, but in fact repetition is frequently a feature, because it makes it easier to refactor the code in the future (less tightly coupled to a particular implementation) and more explicit for the reader.

Reply:
Yeah, I was trying to avoid calling data over and over but if it's fine to do that, that's fine.